File: PS_10_DivideAndConquer/QuickSort.cpp
#include <iostream>
#include <vector>
using namespace std;

void quickSort(vector<int>& nums, int start, int end) {
    if (start >= end) {
        return; // Base case: If the part of the vector has 0 or 1 element, it's already sorted.
    }

    int pivot = nums[(start + end) / 2]; // Using the middle element as the pivot.
    int i = start, j = end;

    while (i <= j) {
        while (nums[i] < pivot) i++; // Move right until an element >= pivot is found.
        while (nums[j] > pivot) j--; // Move left until an element <= pivot is found.

        if (i <= j) {
            swap(nums[i], nums[j]); // Swap the elements.
            i++;
            j--;
        }
    }

    // Recursively sort the two partitions.
    quickSort(nums, start, j);
    quickSort(nums, i, end);
}

int main() {
    vector<int> nums = {10, 5, 17, 9, 6, -10, -39};
    cout << "Before: ";
    for (int i : nums) {
        cout << i << " ";
    }
    cout << endl;

    quickSort(nums, 0, nums.size() - 1);

    cout << "After: ";
    for (int i : nums) {
        cout << i << " ";
    }
    cout << endl;

    return 0;
}

File: PS_11_TwoPointers/QuickSort.java
package myUsacoSilverSolutions.PS_11_TwoPointers;

public class QuickSort {
    static int[] arr = {3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 8, 9, 7, 8, 9};

    public static void quicksort(int[] arr, int start, int end) {
        if (start >= end) return;

        // Partition Logic:
        // 1: We don't need to use extra memory unlike merge sort.
        // 2: To do this we will use two pointers.
        // One for elements smaller than the pivot (j), one for larger (i).

        // Start by choosing a pivot. (I'll use the rightmost element as pivot)
        int pivot = arr[end];
        
        int j = start - 1; // This pointer will help us find elements smaller than the pivot
        // The pointer i, the one for the large will move by default.
        for (int i = start; i < end; i++) {
            if (arr[i] < pivot) {
                j++;
                // Swap arr[j] and arr[i].
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Swap the pivot into the correct position
        int temp = arr[j + 1];
        arr[j + 1] = arr[end];
        arr[end] = temp;

        // Recursive call: Deal with the left and right part of our pivot.
        quicksort(arr, start, j); // Right boundary is `j` because pivot is at `j + 1`
        quicksort(arr, j + 2, end); // Left boundary is `j + 2` because pivot is at `j + 1`
    }

    public static void main(String[] args) {
        quicksort(arr, 0, arr.length - 1);
        for (int i: arr)
            System.out.print(i + " ");
    }
}
